src/contentsign/ClickBounty.sol

```diff
@@ -0,0 +1,286 @@
+        _checkRole(DEFAULT_ADMIN_ROLE);
+        token.safeTransfer(recipient, amount);
+    }
+
+    // -------------------------------------------------------------------------
+    // Oracle Function
+    // -------------------------------------------------------------------------
+
+    /**
+     * @notice Pays a bounty for a given token ID.
+     *         - If `amount < entryFee`, the difference is collected from the token’s owner.
+     *         - If `amount > entryFee`, the surplus is paid out to the owner.
+     *
+     * @dev Only callable by addresses with ORACLE_ROLE.
+     *
+     * Requirements:
+     * - The token must not have been paid before (`bounties[tokenId] == 0`).
+     *
+     * @param tokenId The token ID to which the bounty is awarded.
+     * @param amount  The bounty amount in `token`.
+     *
+     * Emits:
+     * - A transfer event from the token’s owner to this contract (if `amount < entryFee`).
+     * - A transfer event from this contract to the token’s owner (if `amount > entryFee`).
+     * - A {BountyIssued} event upon successful bounty payment.
+     *
+     * Reverts:
+     * - If `bounties[tokenId] != 0` (bounty already paid).
+     */
+    function awardBounty(uint256 tokenId, uint256 amount) external {
+        _checkRole(ORACLE_ROLE);
+        // Ensure the bounty is non-zero
+        if (amount == 0) {
+            revert ZeroBounty(tokenId);
+        }
+        // Ensure this token ID hasn't been paid yet
+        if (bounties[tokenId] != 0) {
+            revert BountyAlreadyPaid(tokenId);
+        }
+
+        // Identify the token owner via the contentSign contract
+        address owner = contentSign.ownerOf(tokenId);
+
+        // If the bounty is smaller than the entry fee, collect the shortfall from the owner
+        if (amount < entryFee) {
+            token.safeTransferFrom(owner, address(this), entryFee - amount);
+        }
+
+        // Record the bounty
+        bounties[tokenId] = amount;
+
+        // If the bounty is larger than the entry fee, pay out the surplus to the owner
+        if (amount > entryFee) {
+            token.safeTransfer(owner, amount - entryFee);
+        }
+
+        emit BountyIssued(tokenId, amount);
+
+        // Update the on-chain leaderboard
+        _updateLeaderboard(tokenId, amount);
+    }
+
+    // -------------------------------------------------------------------------
+    // View Functions
+    // -------------------------------------------------------------------------
+
+    /**
+     * @notice Retrieves an array of `(URI, bountyAmount)` for each token in the top winners list.
+     * @dev The size of the returned array is up to `LEADERBOARD_SIZE`, but might be smaller if
+     *      fewer than `LEADERBOARD_SIZE` bounties have been paid.
+     * @return uriBounties A dynamic array of `URIBounty` structs, each with `uri` and `bounty`.
+     */
+    function getLeaderboard() external view returns (URIBounty[] memory uriBounties) {
```

Comment:
FYI an indexer can do this too. But I understand it might be more convenient here